src/sagemaker_xgboost_container/callback.py

```diff
@@ -1,29 +1,32 @@
 import logging
 from smdebug.xgboost import Hook
+import xgboost as xgb
 
 
 logger = logging.getLogger(__name__)
 
 
-def add_debugging(callbacks, hyperparameters, train_dmatrix,
-                  val_dmatrix=None, json_config_path=None):
-    """Add a sagemaker debug hook to a list of callbacks.
-    :param callbacks: List of callback functions.
+class SmeDebugHook(xgb.callback.TrainingCallback, Hook):
+    """Mix-in callback class. smedebug.xgboost.Hook uses legacy callback style
+    and since XGB-3.0.0 mixing legacy callback instances with new TrainingCallback
+    instances is not allowed.
+
+    See: https://github.com/dmlc/xgboost/blob/v1.3.0/python-package/xgboost/training.py#L92-L93
+
     :param hyperparameters: Dict of hyperparamters.
                             Same as `params` in xgb.train(params, dtrain).
     :param train_dmatrix: Training data set.
     :param val_dmatrix: Validation data set.
-    :param json_config_path: If specified, this json config will be used
-                             instead of default config file.
     """
-    try:
-        hook = Hook.hook_from_config(json_config_path)
-        hook.hyperparameters = hyperparameters
-        hook.train_data = train_dmatrix
-        if val_dmatrix is not None:
-            hook.validation_data = val_dmatrix
-        callbacks.append(hook)
-        logging.info("Debug hook created from config")
-    except Exception as e:
-        logging.debug("Failed to create debug hook", e)
-        return
+    def __init__(self, hyperparameters, train_dmatrix,
+                 val_dmatrix=None):
+        try:
+            Hook.__init__(self)
+            self.hyperparameters = hyperparameters
+            self.train_data = train_dmatrix
+            if val_dmatrix is not None:
+                self.validation_data = val_dmatrix
+            logging.info("Debug hook initialized")
+        except Exception as e:
+            logging.debug("Failed to create debug hook", e)
+            return
```

Comment:
Sorry, I should have caught this initially, but this implementation is actually a bit different from the existing because 

1. we no longer take the `json_config_path` to create the hook 
2. the initialized `SmeDebugHook` will always be appended to `callbacks`, even if an exception is thrown in `__init__` 

Here's what I just tested in ipython:

```python3
class A:
    pass

class B:
    def some_function(self, foo):
        return foo

class MixIn(A, B):
    pass

# MixIn instance has access to inherited function even without calling its parent init. Returns "some input"
MixIn().some_function("some input")
isinstance(MixIn(), A)  # True
isinstance(MixIn(), B)  # Also true
```

Therefore, can you try this instead?

```python3
class SmeDebugHook(xgb.callback.TrainingCallback, Hook):
    """Mix-in callback class. smedebug.xgboost.Hook uses legacy callback style
    and since XGB-3.0.0 mixing legacy callback instances with new TrainingCallback
    instances is not allowed.
    
    See: https://github.com/dmlc/xgboost/blob/v1.3.0/python-package/xgboost/training.py#L92-L93
    """"
    pass

def add_debugging(callbacks, hyperparameters, train_dmatrix,
                  val_dmatrix=None, json_config_path=None):
    """Add a sagemaker debug hook to a list of callbacks.
    :param callbacks: List of callback functions.
    :param hyperparameters: Dict of hyperparamters.
                            Same as `params` in xgb.train(params, dtrain).
    :param train_dmatrix: Training data set.
    :param val_dmatrix: Validation data set.
    :param json_config_path: If specified, this json config will be used
                             instead of default config file.
    """
    try:
        hook = SmeDebugHook.hook_from_config(json_config_path)
        hook.hyperparameters = hyperparameters
        hook.train_data = train_dmatrix
        if val_dmatrix is not None:
            hook.validation_data = val_dmatrix
        logging.info("Debug hook created from config")
    except Exception as e:
        logging.debug("Failed to create debug hook", e)
    else:
        callbacks.append(hook)
```

And then you can just revert the changes in `train.py`, too. 

I think you could also write this if you prefer:

```python3
class SmeDebugHook(xgb.callback.TrainingCallback, Hook):
    """Mix-in callback class. smedebug.xgboost.Hook uses legacy callback style
    and since XGB-3.0.0 mixing legacy callback instances with new TrainingCallback
    instances is not allowed.
    
    See: https://github.com/dmlc/xgboost/blob/v1.3.0/python-package/xgboost/training.py#L92-L93

    :param json_config_path: If specified, this json config will be used
                             instead of default config file.
    :param hyperparameters: Dict of hyperparamters.
                            Same as `params` in xgb.train(params, dtrain).
    :param train_dmatrix: Training data set.
    :param val_dmatrix: Validation data set.
    """"
    def __init__(self, json_config_path, hyperparameters, train_dmatrix, val_dmatrix):
        self = self.hook_from_config(json_config_path)
        self.hyperparameters = hyperparameters
        self.train_data = train_dmatrix
        if val_dmatrix is not None:
            self.validation_data = val_dmatrix

def add_debugging(callbacks, hyperparameters, train_dmatrix,
                  val_dmatrix=None, json_config_path=None):
    """Add a sagemaker debug hook to a list of callbacks.
    :param callbacks: List of callback functions.
    :param hyperparameters: Dict of hyperparamters.
                            Same as `params` in xgb.train(params, dtrain).
    :param train_dmatrix: Training data set.
    :param val_dmatrix: Validation data set.
    :param json_config_path: If specified, this json config will be used
                             instead of default config file.
    """
    try:
        hook = SmeDebugHook(json_config_path, hyperparameters, train_dmatrix, val_dmatrix)
        logging.info("Debug hook created from config")
    except Exception as e:
        logging.debug("Failed to create debug hook", e)
    else:
        callbacks.append(hook)
```

It's the call to `hook_from_config` that will throw if `json_config_path` does not exist: https://github.com/awslabs/sagemaker-debugger/blob/4a470d9c7711e6ff89e5557d91260ee86a0b71b3/smdebug/core/json_config.py#L80 so we still want to make sure we avoid appending the callback if this happens.

Reply:
Oh Cool, Thanks!. Sorry I should've been more careful about that try ... catch .. . I was thinking as we never specify json_path when calling, so try to initialize in a clearer way. I will update in next commit
```
            add_debugging(callbacks=callbacks, hyperparameters=train_cfg, train_dmatrix=train_dmatrix,
                          val_dmatrix=val_dmatrix)
```

Reply:
Oh, that's a good point. 

Can you just do a quick search and see if we ever call `add_debugging` with `json_config_path`? It may be in some notebooks, perhaps, or example scripts. If you can't find any references, then let's initialize using your initial approach since I saw this comment here: https://github.com/awslabs/sagemaker-debugger/blob/4a470d9c7711e6ff89e5557d91260ee86a0b71b3/smdebug/xgboost/hook.py#L156

So we can still have the `add_debugging` function, but in `SmeDebugHook` instead of calling `self = self.hook_from_config(json_config_path)` we would do what you did originally with `Hook.__init__(self)`.

Reply:
I think the reason why it didn't throw FileNotFound Error is becasue there's default json file existing 
https://github.com/awslabs/sagemaker-debugger/blob/4a470d9c7711e6ff89e5557d91260ee86a0b71b3/smdebug/core/json_config.py#L88

> path = os.getenv(CONFIG_FILE_PATH_ENV_STR, DEFAULT_CONFIG_FILE_PATH)

Reply:
This is the original PR: https://github.com/aws/sagemaker-xgboost-container/pull/72 I'll look internally and see if I can find anything pertinent.